### ISSUE TITLE
BUG:  .notdef glyph has to be present in fonts in fontlist

### ISSUE THREAD

**Issue by story645** [matplotlib/matplotlib#27232]
> This also happens with the xkcd plotting style using the "Humor Sans" font.

So uh, this is why #26854 is failing (the plt.xkcd examples seem to work because they don't have a negative in the labels anywhere): throws "glyph missing" warning for title, omits hyphen for ticks:
![Figure_1](https://github.com/matplotlib/matplotlib/assets/1300499/d79564f3-846b-40a3-b4d0-cbbd65b456fc)

_Originally posted by @story645 in https://github.com/matplotlib/matplotlib/issues/17007#issuecomment-1786319109_

per the discussion downthread, if a font does not have a designated [.notdef](https://en.wikipedia.org/wiki/Unicode_input#Availability) then the missing glyph character (the .notdef) is not rendered.

# Suggested Fix
For this one character, we should have a fallback such that we always render some glyph when a glyph is not found. This is particularly because of situations like the missing - at the front of a text, where it may be easy to miss that it's missing.

**Comment by ksunden** [comment]
what would you suggest we do about this?

A glyph missing from a font is not really something we have control over.

Also, I just installed `Humor Sans` locally (from both http://antiyawn.com/uploads/humorsans.html and the debian archives), and it appears to work fine with the negative sign, including in the style reference that is failing in that PR.

This includes explicitly adding the text `"-\N{MINUS SIGN}"`

Further the default formatter uses "-" (`chr(45)` not `chr(8722)`, as far as I can tell, so not sure where that is coming from? are we getting leaking rcParams from another example? 

Perhaps the xkcd style file should explicitly set `:rc:axes.unicode_minus` to `False`, as that may be the source of the unicode character being used. (Though I'll note that locally I _still_ did not get a warning when that rcparam was set to true)

This appears to be `False` by default, but seems related.

https://github.com/matplotlib/matplotlib/blob/b318dbfd80b22709e269b3e7d0cad6e3312a2561/lib/matplotlib/ticker.py#L253

**Comment by tacaswell** [comment]
It looks like we get the tofu from the first font in the fallback list:

https://github.com/matplotlib/matplotlib/blob/cc46ee7bd0cbaa83257d0bc3f808ca42b8c0590e/src/ft2font.cpp#L515-L523

which suggests that the tofu in that font is empty?

Maybe the fix is to use a fallback font with Humor Sans?

**Comment by tacaswell** [comment]
I suspect that in `xkcd()` we can set the font to be something like `['humor sans', 'deja vu']` ?

**Comment by tacaswell** [comment]
I am not convinced that this is something we want to do.

Do we have any evidence that this is a wider problem than that one font? 

If I am reading our code and https://freetype.org/freetype2/docs/tutorial/step2.html, correctly it looks like the ".notdef" glyph is index `0` in freetype's API and in https://github.com/matplotlib/matplotlib/blob/cc46ee7bd0cbaa83257d0bc3f808ca42b8c0590e/src/ft2font.cpp#L704-L714 we will throw c++ exceptions (which I am 99% sure we turn into Python exceptions via the CALL_CPP macro).  So it is not clear to me that the `.notdef` is actually missing here, but rather that it is empty.

Looking at https://freetype.org/freetype2/docs/reference/ft2-character_mapping.html#ft_get_char_index it seems that freetype will always ensure that 0 is the "missing glyph" so it is unclear to me if it is even possible to detect that it is missing via freetype.  I think this also means that freetype as already done "the .notdef character isn't defined, it fallsback to the default .notdef character." for us.  If there default is empty, then I think we need to understand why they made that choice before we try to second-guess it.

**Comment by tacaswell** [comment]
However, it looks like freetype is _already_ ensuring that `.notdef` is available: 
So the questions are:

 - for this font did freetype make the default `.notdef` or does the font actually have an `.notdef` and it is just empty? (I think it is the second making the rest of this moot for this case)
 - if freetype made the default `.notdef` did they pick empty? and if so why?
 - from the freetype API can we tell that any of the above have happened?

A bigger question is a font even a valid font without a `.notdef` (it seems that type42 requires it and TTF at at least requires the 'post' table which in some forms requires `.notdef`)?

-----

Poking at this font in `fontforge` it looks like it does indeed provide `.notdef` it is blank.  If I am understand this correctly, then there is no way we can detect if the `.notdef` glyph is "good" so none of this proposal would actually fix the problem.

At some level, a font can randomize the mapping between the numbers and the glyphs (e.g. it render "1" to look like "7" or something) and there is nothing we can do about that.

----

The minimal solution is we should burn `Humor Sans` out of our code base (both for this reason and because it replaces € with an emoji of Hitler).  It is probably with cleaving that off of the bigger xkcd PR so we can backport it.

**Comment by QuLogic** [comment]
Are we alright with closing this now that #29356 is in?

**Comment by story645** [comment]
Yup, seems to work in xkcd mode (and some other PR fixed the missing - in xkcd), thanks!
